### PR TITLE
Structured stack 

### PR DESCRIPTION
**Pull request by titzer**
This is my first stab at rewriting AstSemantics.md to reflect a structured stack machine architecture.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,9 +10,9 @@ WebAssembly or wasm is a new, portable, size- and load-time-efficient format sui
 
 WebAssembly is currently being designed as an open standard by a [W3C Community Group](https://www.w3.org/community/webassembly/) that includes representatives from all major browsers. *Expect the contents of this repository to be in flux: everything is still under discussion.*
 
-- **WebAssembly is efficient and fast**: The wasm [AST](AstSemantics.md) is designed to be encoded in a size- and load-time-efficient [binary format](BinaryEncoding.md). WebAssembly aims to execute at native speed by taking advantage of [common hardware capabilities](Portability.md#assumptions-for-efficient-execution) available on a wide range of platforms.
+- **WebAssembly is efficient and fast**: The wasm [AST](Semantics.md) is designed to be encoded in a size- and load-time-efficient [binary format](BinaryEncoding.md). WebAssembly aims to execute at native speed by taking advantage of [common hardware capabilities](Portability.md#assumptions-for-efficient-execution) available on a wide range of platforms.
```
Comment (sbc100) [inline README.md]: Maybe just `wasm is designed...`

File: Semantics.md [hunk 2]
```diff
@@ -1,44 +1,35 @@
...
-implementations need not build an actual AST or maintain an actual stack; they
-need only behave [as if](https://en.wikipedia.org/wiki/As-if_rule) they did so.)
-
-This document explains the high-level design of the AST: its types, constructs, and
-semantics. For full details consult [the formal Specification](https://github.com/WebAssembly/spec),
+# Semantics
+
+This document explains the high-level design of WebAssembly code: its types, constructs, and
+semantics.
+WebAssembly code can be considered a *structured stack machine*; a machine where most computations use a stack
```
Comment (sbc100) [inline Semantics.md]: Reflow this paragraph wrapping at 80?  (and below at other places in this file)

File: Semantics.md [hunk 3]
```diff
@@ -1,44 +1,35 @@
...
+In practice, implementations need not maintain an actual value stack, nor actual data structures for control; they
+need only behave [as if](https://en.wikipedia.org/wiki/As-if_rule) they did so.
+For full details consult [the formal Specification](https://github.com/WebAssembly/spec),
 for file-level encoding details consult [Binary Encoding](BinaryEncoding.md),
 and for the human-readable text representation consult [Text Format](TextFormat.md).
 
-Each function body consists of a list of expressions. All expressions and
-operators are typed, with no implicit conversions or overloading rules.
+Each function body consists of a list of instructions which forms an implicit *block*.
+Execution of a instructions proceeds by way of a traditional *program counter* that advances
```
Comment (lukewagner) [inline Semantics.md]: nit: remove "a" in "a instructions"

**Comment by lukewagner** [comment]
lgtm

File: Semantics.md (hunk 3 above)
Comment (titzer) [inline Semantics.md]: Done.

File: Semantics.md [hunk 4]
```diff
@@ -356,28 +346,32 @@ for example, references to a `block`'s label can only occur within the
...
+from outside the loop. This limitation ensures by construction that all control flow graphs
+are well-structured as in high-level languages like Java, JavaScript, Rust and Go.
+Notice that that a `br` to a `block`'s label is  equivalent to a labeled `break` in
+high-level languages; `br` simply breaks out of a `block`.
+
+### Yielding values from control instructions
+
+Executing the `end` of a `block` or `loop` (including implicit blocks such as in `if` or for a function body) has no effect on the stack.
+
+Executing the `end` of the implicit block for a function body is equivalent to a `return`.
```
Comment (titzer) [inline Semantics.md]: Done.

**Comment by titzer** [comment]
With another LGTM I think this can be merged.

File: Semantics.md (hunk 2 above)
Comment (titzer) [inline Semantics.md]: Probably best done as a separate post-pass on the whole file.

File: Semantics.md [hunk 5]
```diff
@@ -330,54 +321,67 @@ global variables will be necessary to allow sharing reference types between
...
 In practice, outer `block`s can be used to place labels for any given branching
-pattern, except for one restriction: one can't branch into the middle of a loop
-from outside it. This restriction ensures all control flow graphs are well-structured
-in the exact sense as in high-level languages like Java, JavaScript, Rust and Go. To
-further see the parallel, note that a `br` to a `block`'s label is functionally
-equivalent to a labeled `break` in high-level languages in that a `br` simply
-breaks out of a `block`.
+pattern, except that the nesting restriction makes it impossible to branch into the middle of a loop
+from outside the loop. This limitation ensures by construction that all control flow graphs
+are well-structured as in high-level languages like Java, JavaScript, Rust and Go.
```
Comment (rossberg-chromium) [inline Semantics.md]: Nit: Not sure if I'd call Rust a high-level language. Maybe a more relative "higher-level" is better?

File: Semantics.md [hunk 6]
```diff
@@ -330,54 +321,67 @@ global variables will be necessary to allow sharing reference types between
...
+pushed back onto the stack, and the program counter is updated to the end of the construct. 
 
-The `nop`, `br`, `br_if`, `br_table`, and `return` constructs do not yield values.
-Other control constructs may yield values if their subexpressions yield values:
+Branches that target a `loop` do not yield a value; they pop any values pushed onto the stack since the start of the loop and set the program counter to the start of the loop.
 
-* `block`: yields either the value of the last expression in the block or the result of an inner branch that targeted the label of the block
-* `loop`: yields the value of the last expression in the loop
-* `if`: yields either the value of the last *then* expression or the last *else* expression or the result of an inner branch that targeted the label of one of these.
+The `drop` operator can be used to explicitly pop and discard values from the stack.
```
Comment (rossberg-chromium) [inline Semantics.md]: s/values/a value/

File: Semantics.md [hunk 7]
```diff
@@ -330,54 +321,67 @@ global variables will be necessary to allow sharing reference types between
...
-from outside it. This restriction ensures all control flow graphs are well-structured
-in the exact sense as in high-level languages like Java, JavaScript, Rust and Go. To
-further see the parallel, note that a `br` to a `block`'s label is functionally
-equivalent to a labeled `break` in high-level languages in that a `br` simply
-breaks out of a `block`.
+pattern, except that the nesting restriction makes it impossible to branch into the middle of a loop
+from outside the loop. This limitation ensures by construction that all control flow graphs
+are well-structured as in high-level languages like Java, JavaScript, Rust and Go.
+Notice that that a branch to a `block`'s label is  equivalent to a labeled `break` in
+high-level languages; branches simply break out of a `block`.
```
Comment (rossberg-chromium) [inline Semantics.md]: Similarly, a branch to a loop is similar to a labeled `continue`.

File: Semantics.md [hunk 8]
```diff
@@ -330,54 +321,67 @@ global variables will be necessary to allow sharing reference types between
...
+
+### Execution semantics of control instructions
```
Comment (rossberg-chromium) [inline Semantics.md]: Note somewhere in this section that in the MVP, block-like constructs, functions, branches and returns may at most use/pop one value.

File: Semantics.md [hunk 9]
```diff
@@ -1,44 +1,35 @@
...
+WebAssembly code can be considered a *structured stack machine*; a machine where most computations use a stack
+of values, but control flow is expressed in structured constructs such as blocks, ifs, and loops.
+In practice, implementations need not maintain an actual value stack, nor actual data structures for control; they
+need only behave [as if](https://en.wikipedia.org/wiki/As-if_rule) they did so.
+For full details consult [the formal Specification](https://github.com/WebAssembly/spec),
 for file-level encoding details consult [Binary Encoding](BinaryEncoding.md),
 and for the human-readable text representation consult [Text Format](TextFormat.md).
 
-Each function body consists of a list of expressions. All expressions and
-operators are typed, with no implicit conversions or overloading rules.
```
Comment (titzer) [inline Semantics.md]: Done.

File: Semantics.md (hunk 5 above)
Comment (titzer) [inline Semantics.md]: Done.

File: Semantics.md (hunk 7 above)
Comment (titzer) [inline Semantics.md]: Done.

File: Semantics.md (hunk 8 above)
Comment (titzer) [inline Semantics.md]: Done.

**Comment by rossberg-chromium** [comment]
lgtm

**Pull request merged**